modules/indexing/src/main/java/org/apache/ignite/internal/processors/query/h2/disk/ExternalResultHashIndex.java

```diff
@@ -0,0 +1,559 @@
+        }
+        catch (IOException e) {
+            U.closeQuiet(this);
+
+            throw new IgniteException("Failed to read query result the from spill idx file. [slot=" + slot + ']', e);
+        }
+    }
+
+    /**
+     * Writes hash map entry to file.
+     *
+     * @param slot Entry slot.
+     * @param hashCode Hash code.
+     * @param addr Row address.
+     */
+    private void writeEntryToIndexFile(long slot, int hashCode, long addr) {
+        try {
+            gotoSlot(slot);
+
+            reusableBuff.clear();
+            reusableBuff.putInt(hashCode);
+
+            // We store addr value incremented by 1 to distinguish 0 address from the empty address field.
+            // On the disk 0 means empty, 1 means zero, y means y-1.
+            reusableBuff.putLong(addr + 1);
+            reusableBuff.flip();
+
+            synchronized (this) {
+                checkCancelled();
+
+                fileIo.writeFully(reusableBuff);
+            }
+        }
+        catch (IOException e) {
+            U.closeQuiet(this);
+
+            throw new IgniteException("Failed to write intermediate query result to the spill file.", e);
+        }
+    }
+
+    /**
+     * Sets arbitrary file position.
+     *
+     * @param slot Position to set.
+     */
+    private void gotoSlot(long slot) {
+        try {
+            synchronized (this) {
+                checkCancelled();
+
+                fileIo.position(slot * Entry.ENTRY_BYTES);
+            }
+        }
+        catch (Exception e) {
+            U.closeQuiet(fileIo);
+
+            throw new IgniteException("Failed to reset the index spill file, slot=" + slot, e);
+        }
+    }
+
+    /**
+     * Checks if the capacity of the hashtable is enough and extends it if needed.
+     */
+    private void ensureCapacity() {
+        if (entriesCnt <= LOAD_FACTOR * cap)
+            return;
+
+        FileIO oldFileIo = fileIo;
+        File oldIdxFile = idxFile;
+
+        long oldSize = cap;
+
+        try {
+            initNewIndexFile(oldSize * 2);
+
+            copyDataFromOldFile(oldFileIo, oldIdxFile, oldSize);
+        }
+        finally {
+            U.closeQuiet(oldFileIo);
+            oldIdxFile.delete();
+        }
+    }
+
+    /**
+     * Copies data from the old file to the extended new one.
+     *
+     * @param oldFile Old file channel.
+     * @param oldIdxFile Old index file.
+     * @param oldSize Old file size in slots.
+     */
+    private void copyDataFromOldFile(FileIO oldFile, File oldIdxFile, long oldSize) {
+        try {
+            entriesCnt = 0;
+
+            oldFile.position(0);
+
+            for (long i = 0; i < oldSize; i++) {
+                Entry e = readEntryFromIndexFile(-1, oldFile);
+
+                if (!e.isRemoved() && !e.isEmpty())
+                    putEntryToFreeSlot(e.hashCode(), e.rowAddress());
+            }
+        }
+        catch (IOException e) {
+            U.closeQuiet(fileIo);
```

Comment:
close() method removes file, but in case of exception fileIO is closed without file deletion.
Why we close fileIO, but keep file on disk?

Reply:
Replaced with `U.closeQuiet(this);`